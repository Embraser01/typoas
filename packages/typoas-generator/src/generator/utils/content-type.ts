import {
  ContentObject,
  ReferenceObject,
  SchemaObject,
} from 'openapi3-ts/oas31';
import { Context } from '../../context';

// This key is a simple hack for the createTypeFromSchema function
// for it to generate a Blob type that does not exist by default in json schema.
export const TYPOAS_BLOB_TYPE_KEY = 'x-typoas-blob-type';

export function getContentTypeSchema(
  content: ContentObject | undefined,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  ctx: Context,
): ReferenceObject | SchemaObject | undefined {
  if (!content) {
    return undefined;
  }
  if (content['application/json']) {
    return content['application/json'].schema;
  }

  if (content['text/plain']) {
    return { type: 'string' };
  }

  if (content['application/octet-stream']) {
    return { [TYPOAS_BLOB_TYPE_KEY]: true };
  }
}
