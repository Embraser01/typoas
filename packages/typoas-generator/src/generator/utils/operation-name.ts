function lowerCase(str: string) {
  return str.toLowerCase();
}

function upperCase(str: string) {
  return str.toUpperCase();
}

/**
 * Convert string to camelCase text.
 */
function camelCase(str: string): string {
  return str
    .replace(/[-./]/g, ' ') //convert all hyphens to spaces
    .replace(/\s[a-z]/g, upperCase) //convert first char of each word to UPPERCASE
    .replace(/\s+/g, '') //remove spaces
    .replace(/^[A-Z]/g, lowerCase); //convert first char to lowercase
}

function pascalCase(str: string): string {
  return camelCase(str).replace(/^[a-z]/, upperCase);
}

export function hasUnsupportedIdentifierChar(key: string): boolean {
  return /[-/.+@]/.test(key);
}

export function sanitizeOperationIdName(op: string): string {
  return camelCase(op);
}

export function sanitizeTypeIdentifier(type: string): string {
  return pascalCase(type);
}
