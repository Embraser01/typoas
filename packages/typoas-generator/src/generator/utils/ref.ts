import { Expression, factory, TypeNode } from 'typescript';

export const IMPORT_RUNTIME = 'r';

export enum ExportedRef {
  HttpMethod = 'HttpMethod',
  BaseServerConfiguration = 'BaseServerConfiguration',
  HttpLibrary = 'HttpLibrary',
  RefResolver = 'RefResolver',
  IsomorphicFetchHttpLibrary = 'IsomorphicFetchHttpLibrary',
  SchemaRefResolver = 'SchemaRefResolver',
  ApiException = 'ApiException',
  AuthProvider = 'AuthProvider',
  BasicAuthConfig = 'BasicAuthConfig',
  BearerAuthConfig = 'BearerAuthConfig',
  BaseFlowConfig = 'BaseFlowConfig',
  SecurityAuthentication = 'SecurityAuthentication',
  ApiKeySecurityAuthentication = 'ApiKeySecurityAuthentication',
  HttpSecurityAuthentication = 'HttpSecurityAuthentication',
  OAuth2SecurityAuthentication = 'OAuth2SecurityAuthentication',
  applyTemplating = 'applyTemplating',
  applyTransforms = 'applyTransforms',
  isCodeInRange = 'isCodeInRange',
  serializeParameter = 'serializeParameter',
}

export function createRuntimeRefProperty(exportedRef: ExportedRef): Expression {
  return factory.createPropertyAccessExpression(
    factory.createIdentifier(IMPORT_RUNTIME),
    exportedRef,
  );
}

export function createRuntimeRefType(
  exportedRef: ExportedRef,
  typeArguments?: readonly TypeNode[],
): TypeNode {
  return factory.createTypeReferenceNode(
    factory.createQualifiedName(
      factory.createIdentifier(IMPORT_RUNTIME),
      factory.createIdentifier(exportedRef),
    ),
    typeArguments,
  );
}

export function createJSONParseWrapper(data: unknown): Expression {
  return factory.createCallExpression(
    factory.createPropertyAccessExpression(
      factory.createIdentifier('JSON'),
      'parse',
    ),
    undefined,
    [factory.createStringLiteral(JSON.stringify(data), true)],
  );
}
