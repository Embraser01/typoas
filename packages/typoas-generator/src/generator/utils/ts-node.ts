import {
  createPrinter,
  EmitHint,
  factory,
  NewLineKind,
  Node,
  NodeFlags,
  SyntaxKind,
} from 'typescript';

export const getStringFromNode = (node: Node): string => {
  const printer = createPrinter({ newLine: NewLineKind.LineFeed });
  const resultFile = factory.createSourceFile(
    [],
    factory.createToken(SyntaxKind.EndOfFileToken),
    NodeFlags.Const,
  );
  return printer.printNode(EmitHint.Unspecified, node, resultFile);
};
