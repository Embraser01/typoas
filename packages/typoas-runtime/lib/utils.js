"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.applyTemplating = exports.serializeParameter = exports.isCodeInRange = void 0;
/**
 * Returns if a specific http code is in a given code range
 * where the code range is defined as a combination of digits
 * and "X" (the letter X) with a length of 3
 *
 * @param codeRange string with length 3 consisting of digits and "X" (the letter X)
 * @param code the http status code to be checked against the code range
 */
function isCodeInRange(codeRange, code) {
    if (codeRange === 'default') {
        return true;
    }
    if (codeRange == code.toString()) {
        return true;
    }
    const codeString = code.toString();
    if (codeString.length != codeRange.length) {
        return false;
    }
    for (let i = 0; i < codeString.length; i++) {
        if (codeRange.charAt(i) != 'X' &&
            codeRange.charAt(i) != codeString.charAt(i)) {
            return false;
        }
    }
    return true;
}
exports.isCodeInRange = isCodeInRange;
/**
 * Serialize a parameter from anything to string.
 */
function serializeParameter(data) {
    if (typeof data === 'string')
        return data;
    if (data instanceof Date)
        return data.toISOString();
    if (typeof data === 'number')
        return data.toString();
    return JSON.stringify(data);
}
exports.serializeParameter = serializeParameter;
function applyTemplating(val, variables) {
    return Object.entries(variables).reduce((url, [key, v]) => url.replace(new RegExp(`{${key}}`, 'g'), encodeURIComponent(serializeParameter(v))), val);
}
exports.applyTemplating = applyTemplating;
