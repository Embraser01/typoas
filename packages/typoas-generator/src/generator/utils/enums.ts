import { SchemaObject } from 'openapi3-ts';
import { factory, EnumMember } from 'typescript';
import { screamingSnakeCase } from './operation-name';

export function canConvertSchemaToEnum(schema: SchemaObject): boolean {
  return (
    schema.type === 'string' &&
    schema.enum !== undefined &&
    schema.enum.every((e) => ![null, '', undefined].includes(e)) &&
    !schema.nullable
  );
}

export function createEnumMembersFromSchema(
  schema: SchemaObject,
): EnumMember[] {
  if (canConvertSchemaToEnum(schema)) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return schema.enum!.map((e) => {
      return factory.createEnumMember(
        factory.createIdentifier(screamingSnakeCase(e)),
        factory.createStringLiteral(e, true),
      );
    });
  }
  throw new Error('Cannot create enum members from this schema');
}
