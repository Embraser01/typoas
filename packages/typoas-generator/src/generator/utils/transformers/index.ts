import {
  ArrayLiteralExpression,
  Expression,
  factory,
  ObjectLiteralElementLike,
  ObjectLiteralExpression,
} from 'typescript';
import { ReferenceObject, SchemaObject } from 'openapi3-ts';
import { Context } from '../../../context';
import { isTransformableLeafDate } from './date';
import { findLeafTransforms, TransformField } from './leaf-transformer-base';
import { createRuntimeRefProperty, ExportedRef } from '../ref';

export function createSchemaTransforms(
  schema: SchemaObject | ReferenceObject,
  ctx: Context,
): ObjectLiteralExpression | null {
  const transforms: ObjectLiteralElementLike[] = [];

  // Date transform
  const rawTransforms = findLeafTransforms(schema, isTransformableLeafDate);
  if (rawTransforms.length) {
    transforms.push(
      factory.createPropertyAssignment(
        'date',
        createFromRawTransforms(rawTransforms, ctx),
      ),
    );
  }

  if (!transforms.length) {
    return null;
  }
  return factory.createObjectLiteralExpression(transforms);
}

function createFromRawTransforms(
  rawTransforms: TransformField[],
  ctx: Context,
): ArrayLiteralExpression {
  return factory.createArrayLiteralExpression(
    rawTransforms.map((transformFields) =>
      factory.createArrayLiteralExpression(
        transformFields.map(([type, data]) => {
          const items: Expression[] = [
            factory.createPropertyAccessExpression(
              createRuntimeRefProperty(ExportedRef.TransformType),
              factory.createIdentifier(type.toUpperCase()),
            ),
          ];

          if (typeof data === 'string') {
            items.push(factory.createStringLiteral(data));
          } else if (Array.isArray(data)) {
            items.push(createFromRawTransforms(data, ctx));
          }

          return factory.createArrayLiteralExpression(items);
        }),
      ),
    ),
  );
}
