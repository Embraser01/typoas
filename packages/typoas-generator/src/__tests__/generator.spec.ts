import { readFileSync } from 'fs';
import { describe, expect, it } from '@jest/globals';
import { resolve } from 'path';
import { OpenAPIObject } from 'openapi3-ts/oas31';
import { getStringFromNode } from '../generator/utils/ts-node';
import { generateClient } from '../index';

describe('create full specs', () => {
  it('should generate client', () => {
    const specs = JSON.parse(
      readFileSync(resolve(__dirname, '../../samples/petstore.json'), 'utf8'),
    ) as OpenAPIObject;

    const node = generateClient(specs, { jsDoc: false });

    expect(getStringFromNode(node)).toMatchSnapshot();
  });

  it('should generate enums if possible', () => {
    const specs = JSON.parse(
      readFileSync(resolve(__dirname, './spec-with-enums.json'), 'utf8'),
    ) as OpenAPIObject;

    const node = generateClient(specs, {
      jsDoc: false,
      generateEnums: true,
    });

    expect(getStringFromNode(node)).toMatchSnapshot();
  });

  it('should allow overrides if possible', () => {
    const specs = JSON.parse(
      readFileSync(resolve(__dirname, './spec-with-overrides.json'), 'utf8'),
    ) as OpenAPIObject;

    const node = generateClient(specs, {
      jsDoc: false,
      generateEnums: true,
      overrides: {
        import: './overrides',
        list: ['PetStatus'],
      },
    });

    expect(getStringFromNode(node)).toMatchSnapshot();
  });

  it('should handle deep refs transforms', () => {
    const specs = JSON.parse(
      readFileSync(resolve(__dirname, './spec-deep-refs.json'), 'utf8'),
    ) as OpenAPIObject;

    const node = generateClient(specs, {
      jsDoc: false,
      generateEnums: true,
    });

    expect(getStringFromNode(node)).toMatchSnapshot();
  });
});
