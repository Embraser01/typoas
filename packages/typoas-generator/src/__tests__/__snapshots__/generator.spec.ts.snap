// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create full specs should generate client 1`] = `
"export class PetStoreClient {
    private authMethods: {
        petstore_auth: r.SecurityAuthentication;
        api_key: r.SecurityAuthentication;
    };
    constructor(private server: r.BaseServerConfiguration, authMethodConfig: AuthMethods, private http: r.HttpLibrary = new r.IsomorphicFetchHttpLibrary, private resolver: r.SchemaRefResolver = new r.RefResolver(JSON.parse('{\\"Order\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"integer\\",\\"format\\":\\"int64\\",\\"example\\":10},\\"petId\\":{\\"type\\":\\"integer\\",\\"format\\":\\"int64\\",\\"example\\":198772},\\"quantity\\":{\\"type\\":\\"integer\\",\\"format\\":\\"int32\\",\\"example\\":7},\\"shipDate\\":{\\"type\\":\\"string\\",\\"format\\":\\"date-time\\"},\\"status\\":{\\"type\\":\\"string\\",\\"description\\":\\"Order Status\\",\\"example\\":\\"approved\\",\\"enum\\":[\\"placed\\",\\"approved\\",\\"delivered\\"]},\\"complete\\":{\\"type\\":\\"boolean\\"}},\\"xml\\":{\\"name\\":\\"order\\"}},\\"Customer\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"integer\\",\\"format\\":\\"int64\\",\\"example\\":100000},\\"username\\":{\\"type\\":\\"string\\",\\"example\\":\\"fehguy\\"},\\"address\\":{\\"type\\":\\"array\\",\\"xml\\":{\\"name\\":\\"addresses\\",\\"wrapped\\":true},\\"items\\":{\\"$ref\\":\\"#/components/schemas/Address\\"}}},\\"xml\\":{\\"name\\":\\"customer\\"}},\\"Address\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"street\\":{\\"type\\":\\"string\\",\\"example\\":\\"437 Lytton\\"},\\"city\\":{\\"type\\":\\"string\\",\\"example\\":\\"Palo Alto\\"},\\"state\\":{\\"type\\":\\"string\\",\\"example\\":\\"CA\\"},\\"zip\\":{\\"type\\":\\"string\\",\\"example\\":\\"94301\\"}},\\"xml\\":{\\"name\\":\\"address\\"}},\\"Category\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"integer\\",\\"format\\":\\"int64\\",\\"example\\":1},\\"name\\":{\\"type\\":\\"string\\",\\"example\\":\\"Dogs\\"}},\\"xml\\":{\\"name\\":\\"category\\"}},\\"User\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"integer\\",\\"format\\":\\"int64\\",\\"example\\":10},\\"username\\":{\\"type\\":\\"string\\",\\"example\\":\\"theUser\\"},\\"firstName\\":{\\"type\\":\\"string\\",\\"example\\":\\"John\\"},\\"lastName\\":{\\"type\\":\\"string\\",\\"example\\":\\"James\\"},\\"email\\":{\\"type\\":\\"string\\",\\"example\\":\\"john@email.com\\"},\\"password\\":{\\"type\\":\\"string\\",\\"example\\":\\"12345\\"},\\"phone\\":{\\"type\\":\\"string\\",\\"example\\":\\"12345\\"},\\"userStatus\\":{\\"type\\":\\"integer\\",\\"description\\":\\"User Status\\",\\"format\\":\\"int32\\",\\"example\\":1}},\\"xml\\":{\\"name\\":\\"user\\"}},\\"Tag\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"integer\\",\\"format\\":\\"int64\\"},\\"name\\":{\\"type\\":\\"string\\"}},\\"xml\\":{\\"name\\":\\"tag\\"}},\\"Pet\\":{\\"required\\":[\\"name\\",\\"photoUrls\\"],\\"type\\":\\"object\\",\\"properties\\":{\\"id\\":{\\"type\\":\\"integer\\",\\"format\\":\\"int64\\",\\"example\\":10},\\"name\\":{\\"type\\":\\"string\\",\\"example\\":\\"doggie\\"},\\"category\\":{\\"$ref\\":\\"#/components/schemas/Category\\"},\\"photoUrls\\":{\\"type\\":\\"array\\",\\"xml\\":{\\"wrapped\\":true},\\"items\\":{\\"type\\":\\"string\\",\\"xml\\":{\\"name\\":\\"photoUrl\\"}}},\\"tags\\":{\\"type\\":\\"array\\",\\"xml\\":{\\"wrapped\\":true},\\"items\\":{\\"$ref\\":\\"#/components/schemas/Tag\\"}},\\"status\\":{\\"type\\":\\"string\\",\\"description\\":\\"pet status in the store\\",\\"enum\\":[\\"available\\",\\"pending\\",\\"sold\\"]}},\\"xml\\":{\\"name\\":\\"pet\\"}},\\"ApiResponse\\":{\\"type\\":\\"object\\",\\"properties\\":{\\"code\\":{\\"type\\":\\"integer\\",\\"format\\":\\"int32\\"},\\"type\\":{\\"type\\":\\"string\\"},\\"message\\":{\\"type\\":\\"string\\"}},\\"xml\\":{\\"name\\":\\"##default\\"}}}'))) { this.authMethods = configureAuthMethods(authMethodConfig); }
    async updatePet(params: {}, body: Pet): Promise<Pet> {
        const requestContext = this.server.makeRequestContext(r.applyTemplating('/pet', params), r.HttpMethod.PUT);
        requestContext.setHeaderParam('Content-Type', 'application/json');
        requestContext.setBody(JSON.stringify(body));
        await this.authMethods.petstore_auth.applySecurityAuthentication(requestContext);
        const response = await this.http.send(requestContext);
        if (r.isCodeInRange('200', response.httpStatusCode))
            return r.applyTransforms(await response.body.json(), JSON.parse('{\\"$ref\\":\\"#/components/schemas/Pet\\"}'), this.resolver);
        if (r.isCodeInRange('400', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        if (r.isCodeInRange('404', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        if (r.isCodeInRange('405', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        throw new r.ApiException<string>(response.httpStatusCode, 'Unknown API Status Code!');
    }
    async addPet(params: {}, body: Pet): Promise<Pet> {
        const requestContext = this.server.makeRequestContext(r.applyTemplating('/pet', params), r.HttpMethod.POST);
        requestContext.setHeaderParam('Content-Type', 'application/json');
        requestContext.setBody(JSON.stringify(body));
        await this.authMethods.petstore_auth.applySecurityAuthentication(requestContext);
        const response = await this.http.send(requestContext);
        if (r.isCodeInRange('200', response.httpStatusCode))
            return r.applyTransforms(await response.body.json(), JSON.parse('{\\"$ref\\":\\"#/components/schemas/Pet\\"}'), this.resolver);
        if (r.isCodeInRange('405', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        throw new r.ApiException<string>(response.httpStatusCode, 'Unknown API Status Code!');
    }
    async findPetsByStatus(params: {
        status?: 'available' | 'pending' | 'sold';
    }): Promise<Pet[]> {
        const requestContext = this.server.makeRequestContext(r.applyTemplating('/pet/findByStatus', params), r.HttpMethod.GET);
        requestContext.setHeaderParam('Content-Type', 'application/json');
        if (params.status !== undefined)
            requestContext.setQueryParam('status', r.serializeParameter(params.status));
        await this.authMethods.petstore_auth.applySecurityAuthentication(requestContext);
        const response = await this.http.send(requestContext);
        if (r.isCodeInRange('200', response.httpStatusCode))
            return r.applyTransforms(await response.body.json(), JSON.parse('{\\"type\\":\\"array\\",\\"items\\":{\\"$ref\\":\\"#/components/schemas/Pet\\"}}'), this.resolver);
        if (r.isCodeInRange('400', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        throw new r.ApiException<string>(response.httpStatusCode, 'Unknown API Status Code!');
    }
    async findPetsByTags(params: {
        tags?: string[];
    }): Promise<Pet[]> {
        const requestContext = this.server.makeRequestContext(r.applyTemplating('/pet/findByTags', params), r.HttpMethod.GET);
        requestContext.setHeaderParam('Content-Type', 'application/json');
        if (params.tags !== undefined)
            requestContext.setQueryParam('tags', r.serializeParameter(params.tags));
        await this.authMethods.petstore_auth.applySecurityAuthentication(requestContext);
        const response = await this.http.send(requestContext);
        if (r.isCodeInRange('200', response.httpStatusCode))
            return r.applyTransforms(await response.body.json(), JSON.parse('{\\"type\\":\\"array\\",\\"items\\":{\\"$ref\\":\\"#/components/schemas/Pet\\"}}'), this.resolver);
        if (r.isCodeInRange('400', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        throw new r.ApiException<string>(response.httpStatusCode, 'Unknown API Status Code!');
    }
    async getPetById(params: {
        petId: number;
    }): Promise<Pet> {
        const requestContext = this.server.makeRequestContext(r.applyTemplating('/pet/{petId}', params), r.HttpMethod.GET);
        requestContext.setHeaderParam('Content-Type', 'application/json');
        await this.authMethods.api_key.applySecurityAuthentication(requestContext);
        await this.authMethods.petstore_auth.applySecurityAuthentication(requestContext);
        const response = await this.http.send(requestContext);
        if (r.isCodeInRange('200', response.httpStatusCode))
            return r.applyTransforms(await response.body.json(), JSON.parse('{\\"$ref\\":\\"#/components/schemas/Pet\\"}'), this.resolver);
        if (r.isCodeInRange('400', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        if (r.isCodeInRange('404', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        throw new r.ApiException<string>(response.httpStatusCode, 'Unknown API Status Code!');
    }
    async updatePetWithForm(params: {
        petId: number;
        name?: string;
        status?: string;
    }): Promise<any> {
        const requestContext = this.server.makeRequestContext(r.applyTemplating('/pet/{petId}', params), r.HttpMethod.POST);
        requestContext.setHeaderParam('Content-Type', 'application/json');
        if (params.name !== undefined)
            requestContext.setQueryParam('name', r.serializeParameter(params.name));
        if (params.status !== undefined)
            requestContext.setQueryParam('status', r.serializeParameter(params.status));
        await this.authMethods.petstore_auth.applySecurityAuthentication(requestContext);
        const response = await this.http.send(requestContext);
        if (r.isCodeInRange('405', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        throw new r.ApiException<string>(response.httpStatusCode, 'Unknown API Status Code!');
    }
    async deletePet(params: {
        api_key?: string;
        petId: number;
    }): Promise<any> {
        const requestContext = this.server.makeRequestContext(r.applyTemplating('/pet/{petId}', params), r.HttpMethod.DELETE);
        requestContext.setHeaderParam('Content-Type', 'application/json');
        if (params.api_key !== undefined)
            requestContext.setHeaderParam('api_key', r.serializeParameter(params.api_key));
        await this.authMethods.petstore_auth.applySecurityAuthentication(requestContext);
        const response = await this.http.send(requestContext);
        if (r.isCodeInRange('400', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        throw new r.ApiException<string>(response.httpStatusCode, 'Unknown API Status Code!');
    }
    async uploadFile(params: {
        petId: number;
        additionalMetadata?: string;
    }, body: any): Promise<ApiResponse> {
        const requestContext = this.server.makeRequestContext(r.applyTemplating('/pet/{petId}/uploadImage', params), r.HttpMethod.POST);
        requestContext.setHeaderParam('Content-Type', 'application/json');
        if (params.additionalMetadata !== undefined)
            requestContext.setQueryParam('additionalMetadata', r.serializeParameter(params.additionalMetadata));
        requestContext.setBody(JSON.stringify(body));
        await this.authMethods.petstore_auth.applySecurityAuthentication(requestContext);
        const response = await this.http.send(requestContext);
        if (r.isCodeInRange('200', response.httpStatusCode))
            return r.applyTransforms(await response.body.json(), JSON.parse('{\\"$ref\\":\\"#/components/schemas/ApiResponse\\"}'), this.resolver);
        throw new r.ApiException<string>(response.httpStatusCode, 'Unknown API Status Code!');
    }
    async getInventory(params: {}): Promise<{
        [key: string]: number;
    }> {
        const requestContext = this.server.makeRequestContext(r.applyTemplating('/store/inventory', params), r.HttpMethod.GET);
        requestContext.setHeaderParam('Content-Type', 'application/json');
        await this.authMethods.api_key.applySecurityAuthentication(requestContext);
        const response = await this.http.send(requestContext);
        if (r.isCodeInRange('200', response.httpStatusCode))
            return r.applyTransforms(await response.body.json(), JSON.parse('{\\"type\\":\\"object\\",\\"additionalProperties\\":{\\"type\\":\\"integer\\",\\"format\\":\\"int32\\"}}'), this.resolver);
        throw new r.ApiException<string>(response.httpStatusCode, 'Unknown API Status Code!');
    }
    async placeOrder(params: {}, body: Order): Promise<Order> {
        const requestContext = this.server.makeRequestContext(r.applyTemplating('/store/order', params), r.HttpMethod.POST);
        requestContext.setHeaderParam('Content-Type', 'application/json');
        requestContext.setBody(JSON.stringify(body));
        const response = await this.http.send(requestContext);
        if (r.isCodeInRange('200', response.httpStatusCode))
            return r.applyTransforms(await response.body.json(), JSON.parse('{\\"$ref\\":\\"#/components/schemas/Order\\"}'), this.resolver);
        if (r.isCodeInRange('405', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        throw new r.ApiException<string>(response.httpStatusCode, 'Unknown API Status Code!');
    }
    async getOrderById(params: {
        orderId: number;
    }): Promise<Order> {
        const requestContext = this.server.makeRequestContext(r.applyTemplating('/store/order/{orderId}', params), r.HttpMethod.GET);
        requestContext.setHeaderParam('Content-Type', 'application/json');
        const response = await this.http.send(requestContext);
        if (r.isCodeInRange('200', response.httpStatusCode))
            return r.applyTransforms(await response.body.json(), JSON.parse('{\\"$ref\\":\\"#/components/schemas/Order\\"}'), this.resolver);
        if (r.isCodeInRange('400', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        if (r.isCodeInRange('404', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        throw new r.ApiException<string>(response.httpStatusCode, 'Unknown API Status Code!');
    }
    async deleteOrder(params: {
        orderId: number;
    }): Promise<any> {
        const requestContext = this.server.makeRequestContext(r.applyTemplating('/store/order/{orderId}', params), r.HttpMethod.DELETE);
        requestContext.setHeaderParam('Content-Type', 'application/json');
        const response = await this.http.send(requestContext);
        if (r.isCodeInRange('400', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        if (r.isCodeInRange('404', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        throw new r.ApiException<string>(response.httpStatusCode, 'Unknown API Status Code!');
    }
    async createUser(params: {}, body: User): Promise<User> {
        const requestContext = this.server.makeRequestContext(r.applyTemplating('/user', params), r.HttpMethod.POST);
        requestContext.setHeaderParam('Content-Type', 'application/json');
        requestContext.setBody(JSON.stringify(body));
        const response = await this.http.send(requestContext);
        if (r.isCodeInRange('default', response.httpStatusCode))
            return r.applyTransforms(await response.body.json(), JSON.parse('{\\"$ref\\":\\"#/components/schemas/User\\"}'), this.resolver);
        throw new r.ApiException<string>(response.httpStatusCode, 'Unknown API Status Code!');
    }
    async createUsersWithListInput(params: {}, body: User[]): Promise<User> {
        const requestContext = this.server.makeRequestContext(r.applyTemplating('/user/createWithList', params), r.HttpMethod.POST);
        requestContext.setHeaderParam('Content-Type', 'application/json');
        requestContext.setBody(JSON.stringify(body));
        const response = await this.http.send(requestContext);
        if (r.isCodeInRange('200', response.httpStatusCode))
            return r.applyTransforms(await response.body.json(), JSON.parse('{\\"$ref\\":\\"#/components/schemas/User\\"}'), this.resolver);
        if (r.isCodeInRange('default', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        throw new r.ApiException<string>(response.httpStatusCode, 'Unknown API Status Code!');
    }
    async loginUser(params: {
        username?: string;
        password?: string;
    }): Promise<string> {
        const requestContext = this.server.makeRequestContext(r.applyTemplating('/user/login', params), r.HttpMethod.GET);
        requestContext.setHeaderParam('Content-Type', 'application/json');
        if (params.username !== undefined)
            requestContext.setQueryParam('username', r.serializeParameter(params.username));
        if (params.password !== undefined)
            requestContext.setQueryParam('password', r.serializeParameter(params.password));
        const response = await this.http.send(requestContext);
        if (r.isCodeInRange('200', response.httpStatusCode))
            return r.applyTransforms(await response.body.json(), JSON.parse('{\\"type\\":\\"string\\"}'), this.resolver);
        if (r.isCodeInRange('400', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        throw new r.ApiException<string>(response.httpStatusCode, 'Unknown API Status Code!');
    }
    async logoutUser(params: {}): Promise<any> {
        const requestContext = this.server.makeRequestContext(r.applyTemplating('/user/logout', params), r.HttpMethod.GET);
        requestContext.setHeaderParam('Content-Type', 'application/json');
        const response = await this.http.send(requestContext);
        if (r.isCodeInRange('default', response.httpStatusCode))
            return r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver);
        throw new r.ApiException<string>(response.httpStatusCode, 'Unknown API Status Code!');
    }
    async getUserByName(params: {
        username: string;
    }): Promise<User> {
        const requestContext = this.server.makeRequestContext(r.applyTemplating('/user/{username}', params), r.HttpMethod.GET);
        requestContext.setHeaderParam('Content-Type', 'application/json');
        const response = await this.http.send(requestContext);
        if (r.isCodeInRange('200', response.httpStatusCode))
            return r.applyTransforms(await response.body.json(), JSON.parse('{\\"$ref\\":\\"#/components/schemas/User\\"}'), this.resolver);
        if (r.isCodeInRange('400', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        if (r.isCodeInRange('404', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        throw new r.ApiException<string>(response.httpStatusCode, 'Unknown API Status Code!');
    }
    async updateUser(params: {
        username: string;
    }, body: User): Promise<any> {
        const requestContext = this.server.makeRequestContext(r.applyTemplating('/user/{username}', params), r.HttpMethod.PUT);
        requestContext.setHeaderParam('Content-Type', 'application/json');
        requestContext.setBody(JSON.stringify(body));
        const response = await this.http.send(requestContext);
        if (r.isCodeInRange('default', response.httpStatusCode))
            return r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver);
        throw new r.ApiException<string>(response.httpStatusCode, 'Unknown API Status Code!');
    }
    async deleteUser(params: {
        username: string;
    }): Promise<any> {
        const requestContext = this.server.makeRequestContext(r.applyTemplating('/user/{username}', params), r.HttpMethod.DELETE);
        requestContext.setHeaderParam('Content-Type', 'application/json');
        const response = await this.http.send(requestContext);
        if (r.isCodeInRange('400', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        if (r.isCodeInRange('404', response.httpStatusCode))
            throw new r.ApiException<any>(response.httpStatusCode, r.applyTransforms(await response.body.json(), JSON.parse('{}'), this.resolver));
        throw new r.ApiException<string>(response.httpStatusCode, 'Unknown API Status Code!');
    }
}"
`;
